### ISSUE TITLE
copy paste not work, clipboard not work properly

### ISSUE THREAD

**Issue by comsyspro** [LizardByte/Sunshine#1539]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

when connecting with moonlight to sunshine i can't copy from the sunshine system to the moonlight system, i can only paste text from moonlight to sunshine with the "Ctrl+Alt+Shift+V" combination.

is this a sunshine problem or a moonlight problem?

on the documentation of moonlight the key combination "Ctrl+Alt+Shift+C" does "Toggle local cursor display in remote desktop mouse mode" and dosen't copy text.
https://github.com/moonlight-stream/moonlight-docs/wiki/Setup-Guide#keyboardmousegamepad-input-options

how to copy text from sunshine system, is this function integrated?

### Expected Behavior

there should be a key combination to copy text from the sunshine system to the moonlight system

### Additional Context

_

### Host Operating System

Windows

### Operating System Version

10

### Architecture

64 bit

### Sunshine commit or version

0.20

### Package

Windows - installer

### GPU Type

Intel

### GPU Model

no relevant info

### GPU Driver/Mesa Version

no relevant info

### Config

```Shell
no relevant info
```

### Apps

```json
no relevant info
```

### Relevant log output

```shell
no relevant info
```

**Comment by cgutman** [comment]
Clipboard support is one-way as of today. You can copy text from your client (using the normal Ctrl/Command+C key combo) and paste it to the host with Ctrl+Alt+Shift+V, but you cannot do the reverse. Bidirectional clipboard sharing is not supported now.

**Comment by comsyspro** [comment]
i will look if i can develop a transport socket for files and clipboard text, but perhaps this is also possible to include in sunshine/moonlight with an extra data transmit channel.

**Comment by ReenigneArcher** [comment]
My 2 cents is that we should not allow copying from host to client as it could be a security issue. Sunshine and Moonlight is not as secure as other RDP solutions which require authentication for every connection. Sunshine and Moonlight only require authentication one time.

**Comment by comsyspro** [comment]
one can use a vpn connectio for sunshine/moonlight or a secure socket connection which runs in background to achieve this.

**Comment by ReenigneArcher** [comment]
You can but it's not required.

**Comment by Nonary** [comment]
As stated by cgutman bi-directional clipboard sharing is not supported now. It is a planned feature for the future, but for the moment this is not considered an issue, but rather a feature request which has already been submitted as a suggestion I believe.

**Comment by comsyspro** [comment]
i have now developed an own copy'n'paste and file transfer implementation which works bidirectional with moonlight/sunshine.

**Comment by comsyspro** [comment]
no, because it's unindependent software which runs in the background.

**Comment by kimjongbing** [comment]
> no, because it's an unindependent software implementation which runs in the background.

do you have plans to share it? parsec is unreliable due to it being x11 and needing to use xwayland, would be good to use sunshine+moonlight and have clipboard access

**Comment by robertpro** [comment]
For anyone needing a workaround for this, you can use this project, I just tested and it works just fine:

https://github.com/labi-le/belphegor

**Comment by ChonSong** [comment]
## Proposal: Bidirectional Clipboard + Android On-Screen Paste

This proposal covers two related gaps:

1. **Host → client clipboard sync** — currently impossible. Copy on the host, paste on the client.
2. **Android on-screen paste** — the existing client→host clipboard requires `Ctrl+Alt+Shift+V`, which is unusable on Android without a physical keyboard.

### Problem

**Current state:**
- Client → host: works via `LiSendUtf8TextEvent` (Moonlight keyboard.cpp:95-127), triggered by `Ctrl+Alt+Shift+V`
- Host → client: **does not exist**. Zero clipboard code in Sunshine (`grep -r clip src/` returns nothing).
- Android: no physical keyboard = no way to trigger paste-to-host. Users need an on-screen button.

**User flow that should work:**
```
[Host copies text] → [Client pastes text]  ← host→client (NEW)
[Client copies text] → [Host pastes text]  ← client→host (exists, but broken on Android)
```

### Research Summary

I reviewed how other remote desktop / streaming apps implement clipboard sharing:

| App | Change Detection | Formats | Loop Prevention | Security |
|-----|-----------------|---------|-----------------|----------|
| **RustDesk** | Observer (Win/macOS/X11) + Polling (Linux files) | Text, HTML, RTF, Images, Files | Owner marker format `"dyn.com.rustdesk.owner"` | Opt-in permission, FUSE for lazy files |
| **Chromium RD** | Observer (Win/macOS), **None (Linux)** | Text only (UTF-8) | Implicit (no self-trigger) | Always-on during session |
| **MS RDP** | Observer + Virtual Channel "CLIPRDR" | Text, Images, Files, Rich Text | Lock/Unlock PDUs + owner tracking | Group Policy, format filtering |
| **Parsec** | Observer (assumed) | Text only | Unknown | No built-in isolation |
| **Steam Remote** | N/A — **no clipboard** | None | N/A | N/A |
| **AnyDesk/TeamViewer** | Observer (assumed) | Text, maybe images | Unknown | Toggle in settings |

**Key design patterns:**

1. **Observer-based detection** (not polling) — Windows: `AddClipboardFormatListener`, macOS: `NSPasteboard`, Linux: `clipboard_master` crate or X11 selection events. Wayland: `wp-primary-selection` or explicit user action.

2. **Owner marker for loop prevention** — RustDesk registers a custom clipboard format `"dyn.com.rustdesk.owner"` with a marker byte (0b01=host, 0b10=client). When reading, if the marker matches the local side, the content is ignored. This is simpler than MS-RDPECLIP's lock/unlock protocol.

3. **Transport via control channel** — Sunshine already has encrypted control messages for rumble, RGB, LEDs. Adding a `Clipboard` message type follows the same pattern.

4. **Opt-in permission** — RustDesk uses `OPTION_ENABLE_CLIPBOARD` per-connection. Sunshine should have a toggle in the web UI and Moonlight should request access.

### Implementation Plan

#### Sunshine Side (host → client)

1. **New control message** in `src/stream.cpp`:
   ```cpp
   struct control_clipboard_t {
     control_header_v2 header;
     uint32_t seq;        // sequence ID for dedup
     uint8_t origin;      // 0 = host, 1 = client
     uint32_t length;     // payload length in bytes
     char text[];         // UTF-8 text (variable length)
   };
   ```
   Add `IDX_SET_CLIPBOARD` message index alongside `IDX_SET_RGB_LED`, `IDX_SET_PLAYER_LEDS`, etc.

2. **Platform clipboard watchers** (`src/platform/{linux,windows,macos}/misc.cpp`):
   - **Windows**: `AddClipboardFormatListener` + `WM_CLIPBOARDUPDATE` — already implemented in [Apollo](https://github.com/ClassicOldSong/Apollo/blob/master/src/platform/windows/misc.cpp#L1859-L1970), can be ported directly
   - **Linux/X11**: `XFixes` extension + `XConvertSelection` (or `clipboard_master` crate like RustDesk)
   - **Linux/Wayland**: `wp-primary-selection` protocol or `wl_clipboard` signals (requires compositor support). Apollo has a [placeholder](https://github.com/ClassicOldSong/Apollo/blob/master/src/platform/linux/misc.cpp#L1066-L1076), needs real impl
   - **macOS**: `NSPasteboard` observer

3. **Loop prevention**: Register a custom clipboard format `"sunshine.host.owner"` with a marker byte. When setting remote clipboard, include marker. When reading clipboard, skip if marker matches local side. (Follows RustDesk's proven approach.)

4. **Compress + debounce**: Large payloads (>1KB) are zstd-compressed. Rapid changes are debounced (min 100ms interval).

5. **Web UI**: Settings → Clipboard Sharing → [ ] Enable (per-client allowlist)

#### Moonlight Side (client → host + Android)

1. **Host→client receiver**: Listen for `IDX_SET_CLIPBOARD` → decode UTF-8 → set client clipboard via platform API (SDL_SetClipboardText on desktop, ClipboardManager on Android)

2. **Android on-screen paste button**: Add a clipboard icon to the on-screen overlay (next to the existing keyboard button). Tapping it:
   - Reads Android system clipboard via `ClipboardManager.getPrimaryClip()`
   - Sends to host via existing `LiSendUtf8TextEvent` path
   - No physical keyboard required

3. **Loop prevention**: When receiving a clipboard update from the remote side, set a flag to ignore the next local clipboard change event (preventing A→B→A→B loops).

### Protocol Sketch

```
ClipboardCapability (negotiated at session start):
  version = 1
  formats = ["text/plain"]
  max_size = 262144

ClipboardSet (sent on clipboard change):
  seq:      u32       // incrementing sequence number
  origin:   u8        // 0 = host, 1 = client
  mime:     string    // "text/plain;charset=utf-8"
  text:     bytes     // UTF-8 payload (max 256 KiB, zstd-compressed if >1KB)

ClipboardAck (optional):
  seq:      u32       // echo back the received seq
  status:   u8        // 0 = ok, 1 = rejected (disabled), 2 = too large
```

### Reuse Existing Code

| Component | Exists in | Reusable? |
|---|---|---|
| Host get/set clipboard (Windows) | [Apollo](https://github.com/ClassicOldSong/Apollo/blob/master/src/platform/windows/misc.cpp#L1859-L1970) | ✅ Port directly — same codebase |
| Host get/set clipboard (Linux) | [Apollo placeholder](https://github.com/ClassicOldSong/Apollo/blob/master/src/platform/linux/misc.cpp#L1066-L1076) | ❌ Needs real impl |
| Client→host transport | [Moonlight keyboard.cpp:95-127](https://github.com/moonlight-stream/moonlight-qt/blob/master/app/streaming/input/keyboard.cpp#L95-L127) | ✅ Already works via `LiSendUtf8TextEvent` |
| Host→client transport | Neither | ❌ Need new `IDX_SET_CLIPBOARD` |
| Client receive + set clipboard | Neither | ❌ Need new handler |

### Why This Is Canonical

- Reuses the existing encrypted control channel — no new ports, no new protocols
- Follows the same pattern as rumble, RGB, and player LED messages
- Minimal new surface area: one message type, one platform abstraction per OS
- Backward compatible: older Moonlight clients ignore the new message type
- Android on-screen button is a natural extension of the existing keyboard overlay
- Loop prevention uses RustDesk's proven owner marker approach (simpler than MS-RDPECLIP)

### Next Steps

Before writing code, I'd like maintainer feedback on:

1. **Is this a feature you'd accept?** (Issue #1539 was closed as "planned", issue #5384 was closed without comment)
2. **Should I open a parallel issue on Moonlight**, or is there a preferred way to coordinate cross-repo features?
3. **Any concerns with the security model** above? (e.g., require per-paste confirmation instead of per-session allowlist)
4. **Scope**: Should the Android on-screen paste button be a separate PR, or bundled with the host→client feature?
5. **Format support**: Text-only for v1, or text + images? (RustDesk supports both; Chromium is text-only)

If the approach looks good, I'll draft PRs for both repos. Estimated effort:
- Sunshine host→client: ~1-2 weeks (Linux impl + new control message + owner marker)
- Moonlight host→client receiver: ~2-3 days
- Android on-screen paste button: ~2-3 days

Happy to adjust scope — even a minimal "host sends clipboard on explicit hotkey" version would be useful, though the full observer model is more convenient.

**Comment by YNNDAA** [comment]
bump
